mempool/src/shared_mempool.rs

```diff
@@ -277,92 +284,80 @@ where
-    // and we don't want to buffer them or let them get reordered. The inbound
-    // direct-send messages are placed in a bounded FuturesUnordered queue and
-    // allowed to execute concurrently. The .buffer_unordered() also correctly
-    // handles back-pressure, so if mempool is slow the back-pressure will
-    // propagate down to network.
-    let f_inbound_network_task = network_events
-        .filter_map(move |network_event| {
-            trace!("SharedMempoolEvent::NetworkEvent::{:?}", network_event);
-            match network_event {
-                Ok(network_event) => match network_event {
-                    Event::NewPeer(peer_id) => {
-                        OP_COUNTERS.inc("smp.event.new_peer");
-                        new_peer(&peer_info, peer_id);
-                        notify_subscribers(
-                            SharedMempoolNotification::PeerStateChange,
-                            &subscribers,
-                        );
-                        future::ready(None)
-                    }
-                    Event::LostPeer(peer_id) => {
-                        OP_COUNTERS.inc("smp.event.lost_peer");
-                        lost_peer(&peer_info, peer_id);
-                        notify_subscribers(
-                            SharedMempoolNotification::PeerStateChange,
-                            &subscribers,
-                        );
-                        future::ready(None)
-                    }
-                    // Pass through messages to next combinator
-                    Event::Message((peer_id, msg)) => future::ready(Some((peer_id, msg))),
-                    _ => {
-                        security_log(SecurityEvent::InvalidNetworkEventMP)
-                            .error("UnexpectedNetworkEvent")
-                            .data(&network_event)
-                            .log();
-                        unreachable!("Unexpected network event")
+    let mut workers_available = smp.config.shared_mempool_max_concurrent_inbound_syncs;
+    let (sender, mut receiver) = mpsc::channel(workers_available);
+
+    while let Some(event) = network_events.next().await {
+        trace!("SharedMempoolEvent::NetworkEvent::{:?}", event);
+        match event {
+            Ok(network_event) => match network_event {
+                Event::NewPeer(peer_id) => {
+                    OP_COUNTERS.inc("smp.event.new_peer");
+                    new_peer(&peer_info, peer_id);
+                    notify_subscribers(SharedMempoolNotification::PeerStateChange, &subscribers);
+                }
+                Event::LostPeer(peer_id) => {
+                    OP_COUNTERS.inc("smp.event.lost_peer");
+                    lost_peer(&peer_info, peer_id);
+                    notify_subscribers(SharedMempoolNotification::PeerStateChange, &subscribers);
+                }
+                Event::Message((peer_id, mut msg)) => {
+                    if workers_available == 0 {
```

Comment:
a comment here would be nice, since I think the control flow is a bit tricky. like "The first `workers_available` number of workers are issued a free work permit, which they must then release to the pool after they're finished. The subsequent workers must then acquire their work permits from the pool."